### ISSUE TITLE
Move global configuration

### ISSUE THREAD

**Issue by sandstrom** [procore-oss/blueprinter#403]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

If blueprinter is used for different things in a single codebase, the global configuration cannot be used.

For example, one might have two APIs (public and internal), and want to use blueprinter for both, but with different configuration for each.

### Describe the feature you'd like to see implemented

Deprecate global configuration and move configuration to the base class.

Make it possible to override in subclasses, thereby relying on the common class inheritance pattern to handle "global configuration".

```ruby
# One group of blueprinters
class PublicApiBlueprint < Blueprinter::Base
  # DSL alt 1
  config do
    sort_fields_by = :definition
  end

  # DSL alt 2
  config.sort_fields_by = :definition
end

class UserBlueprint < PublicApiBlueprint
  identifier :uuid

  fields :first_name, :last_name
end


# Another group of blueprinters
class InternalApiBlueprint < Blueprinter::Base
  # DSL alt 1
  config do
    sort_fields_by = :score
  end

  # DSL alt 2
  config.sort_fields_by = :score
end

class GroupBlueprint < InternalApiBlueprint
  identifier :id

  fields :name, :members_count
end
```

As for making sure configs are not shared (only copied) down the inheritance hiearchy, I'd probably pull in `class_attribute` from ActiveSupport (https://guides.rubyonrails.org/active_support_core_extensions.html#class-attributes).

If you want to add it manually there is some inspiration here: 
- https://github.com/aws/aws-sdk-ruby-record/pull/124/files#diff-c472e379a508904c94f0edaf4c5094a32ec30d00014575d8b8fe910a4bce8460
- https://stackoverflow.com/questions/1251352/ruby-inherit-code-that-works-with-class-variables (this one doesn't go the whole way though)

**Comment by sandstrom** [comment]
Happy to discuss this in more detail, and elaborate more.

**Comment by lessthanjacob** [comment]
@sandstrom Apologies for the delay here, and thanks for the detailed suggestion!

I think this would be a valuable addition to the library, and leveraging `class_attribute` for part of the implementation seems reasonable to me!

**Comment by ritikesh** [comment]
`class_attribute` is a rails implementation and historically, we've shied away from doing rails specific implementations. We can however, use class instance variables.

**Comment by lessthanjacob** [comment]
> class_attribute is a rails implementation and historically, we've shied away from doing rails specific implementations. We can however, use class instance variables.

Not to split hairs here, but I wouldn't consider it _explicitly_ a Rails implementation; this functionality was extracted from `rails` and is encapsulated entirely within an `activesupport` gem which can technically be used in any Ruby project. 

If we don't want to bring in an intentional dependency on `activesupport` due to the side-effects it has on Ruby core classes, then I think that's a reasonable argument. The library _is_ available to us currently due to our dependency on `activerecord` (which is partly why I wouldn't necessary be opposed to using it here), but truth be told I'd also like to fully remove that dependency as well.

**Comment by sandstrom** [comment]
I don't have a strong preference. 

Should be fairly easy to replicate the logic of `class_attribute` from Rails, using class instance variables and the `inherited` [method](https://ruby-doc.org/3.2.2/Class.html#method-i-inherited).

**Comment by ritikesh** [comment]
This is a major refactor and I would personally prefer combining this with any/some other major deprecations planned for the near future as a potential 2.x release. Maybe this could go into a separate 2.x branch.

**Comment by sandstrom** [comment]
Still relevant!

**Comment by jhollinger** [comment]
I think I'm on board with this. "Global" config could go into an `ApplicationBlueprint` which all others _could_ inherit from.

I'd vote against using ActiveSupport for `class_attribute`. Pretty easy to accomplish the same thing using `inherited`.

**Comment by sandstrom** [comment]
@jhollinger Do you know of a blog post or similar explaining how to replicate `class_attribute` behavior (more or less) with `inherited`? 

Or are you simply thinking of something like this:

```ruby
class Mother
  @things = []

  class << self
    attr_accessor :things
  end  

  def self.inherited(subclass)
    puts "New subclass: #{subclass}"
    subclass.things = self.things.dup
  end
end

class Child < Mother
end
```

**Comment by jhollinger** [comment]
@sandstrom Yes, assuming we only have a handful of attributes to keep track of, the above is what I was thinking.

**Comment by RazvanFarte** [comment]
Hi guys,

Hope I'm on the right topic. Should this mean that after this feature is implemented, it will be possible to have two separate `sort_fields_by` configs, per class level?

I'm trying to work around an issue with setting this configs, that impacts the views and associations in another Blueprint.
The issue happens due to this line `sorted_fields = sort_by_definition ? sort_by_def(view_name, fields) : fields.values.sort_by(&:name)` in `fields_for`.
And the only fix found was to have the configure block before and after the first Blueprint definition, which looks ugly to me. Would prefer sandstorm's version.  Not sure if it rings a bell, but I could open a new issue, so it could be referenced here and hopefully some configs to be available at Blueprint level, instead of global.

Here the fix I found that works for me:
`# frozen_string_literal: true

# TODO. Find a better way to apply this config
Blueprinter.configure do |config|
  config.sort_fields_by = :definition
end

module CustomerApi
  class BaseBlueprint < Blueprinter::Base
  end
end

Blueprinter.configure do |config|
  config.sort_fields_by = :name_asc
end
`

**Comment by sandstrom** [comment]
> Should this mean that after this feature is implemented…

Yes, that would be possible

**Comment by sandstrom** [comment]
Not stale

**Comment by jhollinger** [comment]
FYI, this is now part of a proposed V2 base class in #437 (along with other changes). All configuration, and fields, are held in Blueprint classes and are inheritable. To simulate "global" config in V2, options and default fields can be configured in an `ApplicationBlueprint` class which others inherit from.

**Comment by sandstrom** [comment]
Not stale

**Comment by sandstrom** [comment]
not stale

**Comment by lessthanjacob** [comment]
@sandstrom apologies for the noise with our (fairly aggressive) workflow for stale issues. This is applied broadly across the organization, but I'll look into tweaking it a bit.
